// furrybotapi/build/src/util/get.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const https_1 = __importDefault(require("https"));
const url_1 = __importDefault(require("url"));
async function get(url, ua, auth) {
    return new Promise((a, b) => {
        const u = url_1.default.parse(url);
        https_1.default
            .get({
            host: u.host,
            path: u.path,
            port: url.indexOf("https") !== -1 ? 443 : 80,
            method: "GET",
            headers: Object.assign({ "User-Agent": ua, "Host": "api.furry.bot" }, (!!auth ? ({
                Authorization: auth
            }) : ({})))
        }, (res) => {
            const d = [];
            res
                .on("data", d.push.bind(d))
                .on("error", b)
                .on("end", () => a({
                headers: res.headers,
                body: Buffer.concat(d),
                statusCode: res.statusCode,
                statusMessage: res.statusMessage
            }));
        })
            .end();
    });
}
// for overrides
exports.default = get;
