// furrybotapi/build/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Constants_1 = require("./util/Constants");
const Animals_1 = __importDefault(require("./Animals"));
const Furry_1 = __importDefault(require("./Furry"));
const E621_1 = __importDefault(require("./E621"));
class FurryBotAPI {
    constructor(d) {
        if (!d)
            d = {};
        this.options = {
            userAgent: d.userAgent || Constants_1.USER_AGENT,
            apiKey: d.apiKey || "",
            debug: d.debug || (() => null),
            baseURL: d.baseURL || Constants_1.API_URL
        };
    }
    get animals() { return new Animals_1.default(this.options); }
    get furry() { return new Furry_1.default(this.options); }
    get e621() { return new E621_1.default(this.options); }
    get debug() { return this.options.debug; }
}
exports.default = FurryBotAPI;
